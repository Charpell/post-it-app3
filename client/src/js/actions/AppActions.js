import AppDispatcher from '../dispatcher/AppDispatcher';
import AppConstants from '../constants/AppConstants';


const AppActions = {
/**
   * @description describes an action that makes
   * API call to the server for a post request
   * to register a user
   *
   * @param {Object} userDetails an object that contains the username,
   * password, email and phone number of a new user
   *
   * @returns {void} void
   *
   */
  registerUser(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SIGN_UP,
      userDetails
    });
  },

/**
   * @description describes an action that makes
   * API call to the server for a post request
   * to receive a user details
   *
   * @param {Object} users an object containing the user details
   *
   *
   * @returns {void} void
   *
   * @returns {Object} returns registered user registration details
   */
  receiveUsers(users) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_USERS,
      users
    });
  },


/**
   * @description describes an action that makes
   * API call to the server for a post request to get all users
   * who have seen a message.
   *
   * @param {Object} user contains the user and the message Id of a
   * viewed message.
   *
   * @returns {Object} returns an object containing list of users
   */
  seenMessage(user) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SEEN_MESSAGE,
      user
    });
  },

/**
   * @description describes an action that makes
   *  API call to the server for a post request to save a user's group
   *
   * @param {Object} group contains the groupname and the user
   *
   * @returns {Object} returns notification message
   */
  createGroup(group) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.CREATE_GROUP,
      group
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all the groups that a user belongs to.
   *
   * @param {Object} userName the user
   *
   * @returns {void} void
   */
  getGroups(userName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GET_GROUPS,
      userName
    });
  },


  /**
   * @description describes an action that
   * receive all user's notification.
   *
   * @param {Object} notification an object of notifications
   *
   * @returns {Object} returns an object containing user's notification
   */
  receiveNotification(notification) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_NOTIFICATION,
      notification
    });
  },

/**
   * @description describes an action that
   * receive all groups that a user belong to.
   *
   * @param {Object} groups an object of groups
   *
   * @returns {Object} returns an object containing user's group
   */
  receiveGroups(groups) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_GROUPS,
      groups
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to save a user into a group.
   *
   * @param {Object} userDetails the groupname and the username
   *
   * @returns {Object} returns a notification message
   */
  addUserToGroup(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.ADD_USER_TO_GROUP,
      userDetails
    });
  },

  /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to save a message.
   *
   * @param {Object} message an object that contains the group, message,
   * time it was posted, the user who posted the message
   *
   * @returns {void} void
   */
  postMessage(message) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.POST_MESSAGE,
      message
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to receive message.
   *
   * @param {Object} message an object of messages
   *
   * @returns {void} void
   */
  receiveMessages(message) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_MESSAGE,
      message
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to login a user.
   *
   * @param {Object} userDetails an object that contains the email and password
   * of a registered user
   *
   * @returns {void} void
   */
  loginUser(userDetails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SIGN_IN,
      userDetails
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to recieve a user's details.
   *
   * @param {Object} user an object containing the user details
   *
   * @returns {Object} returns registered user details
   */
  receiveLogin(user) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_LOGIN,
      user
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to get users and messages in a group.
   *
   * @param {String} group group name
   *
   * @returns {Object} returns users and messages in a group
   */
  searchUserMessage(group) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.SEARCH_USER_MESSAGE,
      group
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to login a user with a google account
   *
   * @param {Object} googleUser an object that contains the username,
   * email and phone number of a new user
   *
   * @returns {Object} returns user's google account details
   */
  googleLogin(googleUser) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GOOGLE_LOGIN,
      googleUser
    });
  },

 /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to register a user with a google account
   *
   * @param {Object} googleUser an object that contains the username,
   * email and phone number of a new user
   *
   * @returns {Object} returns user's google account details
   */
  googleSignup(googleUser) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.GOOGLE_SIGNUP,
      googleUser
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server to sign the user out
   *
   * @returns {Object} returns registered user details
   */
  logout() {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.LOGOUT
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all users.
   *
   * @param {Object} users object containing list of users
   *
   * @returns {Object} returns an object containing list of users
   */
  receiveUser(users) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_USER,
      users
    });
  },

  /**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all emails.
   *
   * @param {Object} emails object containing list of emails
   *
   * @returns {Object} returns an object containing list of emails
   */
  receiveEmails(emails) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_EMAILS,
      emails
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get all numbers.
   *
   * @param {Object} numbers object containing list of numbers
   *
   * @returns {Object} returns an object containing list of numbers
   */
  receiveNumber(numbers) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RECEIVE_NUMBERS,
      numbers
    });
  },

/**
   * @description describes an action that gets
   * the current user from local storage
   *
   * @param {String} displayName dislplay name of the user
   *
   * @returns { String } returns a displayName
   */
  displayName(displayName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.DISPLAY_NAME,
      displayName
    });
  },

 /**
   * @description describes an action that makes
   * an API call to the server for a post request
   * to reset a user's password.
   *
   * @param {String} email the email address of the user
   *
   * @returns {Object} returns a notification message
   */
  resetPassword(email) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.RESET_PASSWORD,
      email
    });
  },

/**
   * @description describes an action that makes
   * an API call to the server for a get request
   * to get a user's notification.
   *
   * @param {String} userName the user's name
   *
   * @returns {Object} returns an object containing user's notificaions
   */
  getNotification(userName) {
    AppDispatcher.handleViewAction({
      actionType: AppConstants.NOTIFICATIONS,
      userName
    });
  },

};
export default AppActions;
