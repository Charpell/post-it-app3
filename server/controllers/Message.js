import moment from 'moment';
import config from './../config';
import { sendNotification, sendEmailNotification, sendSMSNotification }
from '../helpers/utils';

const { usersRef, groupRef, firebase } = config;

/**
  * @description: A class that controls all message routes
  *
  * @class
  */
class Message {
  /**
 * @description: creates a message and post it through
 *
 * @param {Object} req request object
 * @param {Object} res response object
 *
 * @return {Object} response containing the posted message
 */
  static createMessage(req, res) {
    const { group, user, message, notification, priority } = req.body;

    const currentUser = firebase.auth().currentUser;
    let googleAuth = false;

    usersRef.child(user).child('google').once('value', (snapshot) => {
      if (snapshot.exists()) {
        googleAuth = true;
      }
      if (currentUser || googleAuth) {
        const messageKey = groupRef.child(group).child('Messages').push(
          { user,
            message,
            time: moment().format('h:mm a, MMM Do'),
            priority
          }).key;
        groupRef.child(group).child('Messages').child(messageKey).child('Seen')
        .child('Bot')
        .set('Bot')
        .then(() => {
          res.status(201).json({
            message: 'Message posted successfully',
            messageData: message,
            group,
            user,
            notification,
            priority
          });
        })
        .catch(() => {
          res.status(500).json(
            { message: 'Internal server error' }
          );
        });
        sendNotification(group, user, notification);
        sendEmailNotification(group, priority);
        sendSMSNotification(group, priority);
      } else {
        res.status(401).send('Access denied; You need to sign in');
      }
    });
  }


    /**
 * @description: retrieves all users who have seen a message
 *
 * @param {Object} req request object
 * @param {Object} res response object
 *
 * @return {Object} response containing all users who have seen a message
 */
  static getReadMessageUsers(req, res) {
    const { groupName, messageID } = req.params;
    groupRef.child(groupName)
    .child('Messages')
    .child(messageID).child('Seen')
    .once('value', (users) => {
      res.status(200).json({
        message: 'Users in Group Sent',
        users,
        groupName,
        messageID
      });
    })
    .catch(() => {
      res.status(500).json({
        message: 'Internal server error'
      });
    });
  }

}


export default Message;

