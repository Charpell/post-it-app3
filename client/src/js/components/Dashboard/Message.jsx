import React, { Component } from 'react'
import { Modal } from 'react-bootstrap'

import AppActions from '../../actions/AppActions'
import AppStore from '../../stores/AppStore'

/**
 * @description This displays the message from the database
 * 
 * @class Message
 * 
 * @extends {Component}
 */
export default class Message extends Component {
  constructor(props) {
    super(props);
    this.state = {
      seenMessage: AppStore.getSeenUsers(),
      showModal: false,
    };
    this.onChange = this.onChange.bind(this)
    this.handleSeenMessage = this.handleSeenMessage.bind(this);
    this.handleClose = this.handleClose.bind(this)
  }



  /**
     * @description Makes an action call to tick users who have seen a message
     * 
     * @param {object} event
     * 
     * @returns {void}
     * 
     * @memberof Messsage
  */
  handleSeenMessage(event) {
    event.preventDefault();

    const user = {
      groupName: this.props.group,
      messageID: this.props.message.id
    }
    AppActions.seenMessage(user)
    this.setState({ showModal: true })
  }

  handleClose() {
    this.setState({ showModal: false });

  }

  /**
   * @method componentWillMount
   * 
   * @description Adds an event Listener to the Store and fires when the component is fully mounted.
   * 
   * @return {void}
   * 
   * @memberof Message
   */
  componentWillMount() {
    AppStore.addChangeListener(this.onChange);
  }

  /**
  * @method componentWillUnmount

  * @description Removes event Listener from the Store

  * @memberof Message
  */
  componentWillUnmount() {
    AppStore.removeChangeListener(this.onChange);
  }

  /**
   * @method onChange
   * 
   * @description Monitors changes in the components and change the state
   * 
   * @memberof Message
   */
  onChange() {
    this.setState({ seenMessage: AppStore.getSeenUsers() });
  }

  /**
   * @method render
   * 
   * @description Render react component
   * 
   * @memberof Message
   */
  render() {
    return (
      <div>
          <a href="#" className="list-group-item list-group-item-action flex-column align-items-start">
            <div className="d-flex w-100 justify-content-between">
              <h5 className="mb-1"><strong>{this.props.message.user}</strong>&nbsp;&nbsp; 
                <small className="text-muted">{this.props.message.Time}</small>&nbsp;&nbsp;                 
                <span onClick={this.handleSeenMessage} >  
                  <span className="glyphicon glyphicon-user"></span>
                </span>
              </h5>
            </div>
            <p className="mb-1">{this.props.message.text}</p>
          </a>


        <Modal show={this.state.showModal} onHide={this.handleClose}>
          <Modal.Header closeButton>
            <Modal.Title>Users who have Read Message</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <ul className='mylist'>
            
              {this.state.seenMessage ?
                Object.keys(this.state.seenMessage).map(function (keyName, keyIndex) {
                  return <li key={keyIndex}>{keyName}</li>
                }):
                <li className='mylist'>Empty</li>
              }
            </ul>
          </Modal.Body>
          <Modal.Footer>
            <a href="#/dashboard" onClick={this.handleClose}> Close</a>
          </Modal.Footer>
        </Modal>

      </div>
    )
  }

}
