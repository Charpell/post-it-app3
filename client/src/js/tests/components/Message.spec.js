import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Message from '../../components/Dashboard/Message'
import AppActions from '../../actions/AppActions'

jest.mock('../../../../../server/config', () => ({
  }));
jest.mock('../../actions/AppActions');

let spyOnDispatcher;
beforeEach(() => {
spyOnDispatcher = jest.spyOn(AppActions, 'resetPassword');
});

afterEach(() => {
spyOnDispatcher.mockReset();
});

describe('Message component', () => {
  it('It should render Message component', () => {
    const tree = renderer.create(<Message />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('It should display the necessary elements', () => {
    const wrapper = shallow(<Message />);
    wrapper.instance().componentWillMount();
    wrapper.instance().componentWillUnmount();
    expect(wrapper.find('div').length).toBe(6);
    expect(wrapper.find('h2').length).toBe(1);
    expect(wrapper.find('h4').length).toBe(1);
    expect(wrapper.find('form').length).toBe(1);
    expect(wrapper.find('input').length).toBe(1);
    expect(wrapper.find('a').length).toBe(1);
    expect(wrapper.find('br').length).toBe(3);
});

  it('It should expect handleSeenMessage AppAction to be called', () => {
    const spyOnDispatcher = spyOn(AppActions, 'seenMessage');
    const event = {
        target: {
          name: 'name',
          value: 'value',
        },
        preventDefault: () => jest.fn()
      };
    const wrapper = mount(<Message />);
    wrapper.instance().refs.email.value = 'someemail@email.com';
    wrapper.instance().handleSeenMessage(event);
    expect(spyOnDispatcher).toHaveBeenCalled();
});

});

