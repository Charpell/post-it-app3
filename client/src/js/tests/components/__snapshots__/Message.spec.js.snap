// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component should create a snapshot of itself 1`] = `
<div>
  <a
    className="list-group-item list-group-item-action flex-column align-items-start"
    href="#"
  >
    <div
      className="d-flex w-100 justify-content-between"
    >
      <h5
        className="mb-1"
      >
        <strong />
          
        <small
          className="text-muted"
        />
          
        <span
          onClick={[Function]}
        >
          <span
            className="glyphicon glyphicon-user"
          />
        </span>
      </h5>
    </div>
    <p
      className="mb-1"
    />
  </a>
</div>
`;
